### PR TITLE
[java] Deprecated CommentUtil, move implementation to AST Comment

### LINKED ISSUE #1174: [java] CommentUtil.multiLinesIn() could lead to StringIndexOutOfBoundsException
pmd/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/CommentUtil.java 

Affects PMD Version: latest version

Description: I'm working to develop my own comment rule to find all code in comment , then I use CommentUtils. I tried to multiline the comment ,but found StringIndexOutOfBoundsException when testing by a test project. actually, it happens in line 108,when it goes to charAt(0)
```
if (line.charAt(0) == '*') {
                filteredLines.add(line.substring(1));
                continue;
}
```
because sometime we make a comment like this
```
/*
    sometime we use ctrl+shift+/ to make a multiLineComment(not use by /** )
     <--- here, thie line causes a string ""        
 */
```
In line 3,when using  CommentUtil.multiLinesIn(),this variable string **line = ""**,cause there is nothing!
so "".charAt(0) throws StringIndexOutOfBoundsException.
In my code,I add a StringUtil.isEmpty(line) to check it. In my comment rule, every empty string is also useful, because I use it to count beginLineNumber. I don't whether PMD provides a better way to multiLine the multiComment or JavaDocComment? I only find this CommentUtil useful, so I hope you guys to add a empty check beyond line 96 just like following code:
 ```
String line = rawLine.trim();
            
            if(StringUtil.isEmpty(line)){
            	continue;
            }
            
            if (line.startsWith("//")) {
                filteredLines.add(line.substring(2));
                continue;
            }
     ...
```
or we can use if("".equals(line)) to check it.

### PR DESCRIPTION
**Pull request by adangel**
Fixes #1174

Work in progress - this is missing:

- [x] Mark the methods in CommentUtil as deprecated
- [x] Move the methods maybe into the Comment AST nodes
- [x] Release Notes (API changes)

### CODE REVIEW

**Comment by adangel** [comment]
That was a interesting rabbit whole :smile: 
Most methods of CommentUtil didn't work (at least, not as I expected). The code has also been duplicated into AbstractCommentRule (e.g. https://github.com/pmd/pmd/blob/a89d5413ab7c17e62a3b26085cca144b3614542d/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/documentation/AbstractCommentRule.java#L96 - looks familiar - this even had the bug already fixed...)

So interesting is: FormalComment has basic support for javadoc tags: In [Comment](https://github.com/pmd/pmd/blob/a89d5413ab7c17e62a3b26085cca144b3614542d/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java#L22) we parse the javadoc and add [JavadocElement](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/JavadocElement.java) nodes as children... I moved the code into FormalComment.
It's of course far away from being complete, but could be a interesting way of exposing the javadoc information.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java [hunk 1]
```diff
@@ -5,45 +5,102 @@
...
+    /**
+     * Removes the leading comment marker (like {@code *}) of each line
+     * of the comment as well as the start marker ({@code //}, {@code /*} or {@code /**}
+     * and the end markers (<code>&#x2a;/</code>).
+     *
+     * @param comment the raw comment
+     * @return List of lines of the comments
+     */
+    private List<String> multiLinesIn() {
+        String[] lines = getImage().split("\\R");
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java]: support for the line break matcher is Java8+ only. We can however use the extended version of this:

`\u000D\u000A|[\u000A\u000B\u000C\u000D\u0085\u2028\u2029]`

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java [hunk 2]
```diff
@@ -5,45 +5,102 @@
...
+        if (lines == null) {
+            return Collections.emptyList();
+        }
+
+        List<String> result = new ArrayList<>(lines.size());
+        List<String> tempList = new ArrayList<>();
+        boolean foundFirstNonEmptyLine = false;
+        for (String line : lines) {
+            if (StringUtils.isNoneBlank(line)) {
+                // new non-empty line: add all previous empty lines occurred before
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java]: `isNoneBlank` doesn't check the line is not empty, but actually if not a single character is a whitespace. So, "this is my comment" will fail this check for having 3 whitespaces. Is this really what you expect?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/FormalComment.java [hunk 3]
```diff
@@ -4,15 +4,40 @@
...
 
     @Override
     public String getXPathNodeName() {
         return "FormalComment";
     }
+
+    private void findJavadocs(String commentText) {
+        Collection<JavadocElement> kids = new ArrayList<>();
+
+        Map<String, Integer> tags = CommentUtil.javadocTagsIn(commentText);
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/FormalComment.java]: shouldn't we avoid internal usage of `CommentUtil` as we are deprecating it?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java (hunk 1 above)
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java]: Oh, didn't know that...

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java (hunk 2 above)
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Comment.java]: I guess, I wanted to use `isNotBlank` - isNoneBlank is the same, but for multiple lines (it accepts a vararg ... String). So, the plural here would refer to multiple lines, not to multiple characters...

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/FormalComment.java (hunk 3 above)
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/FormalComment.java]: I've missed that...

**Pull request merged**